### PR TITLE
claim for airdrop

### PR DESCRIPTION
**Pull request by dylenfu**
implement claim contract for airdrop:
1.contract owner:
- upload user actions merkle root;
- upload the number of each user action;
- transfer enough token to the contract, `_checkBalance` will execute only once if balance sufficient;

2.user prove the action:
- execute `claim` or `withdraw` with parameters of `request{Action action, bytes32[] proofs, byte32 leaf}` which used to prove the user action, and user will get satisfactory reward.

3.user claim for `staking` airdrop, the amount should equals to:
- fig balance: 
  500w / fig_balance_user_count;

- fit governance: 
  60w/ fit_governance_user_count;
  
- discord phase2: 
  90w/discord_phase2_user_count;

- discord level5: 
  150w/ discord_level5_user_count;

4.miner withdraw for `borrow` airdrop, and the token will be released in 6 months, the amount miner can withdraw:
- miner only  borrow: 
   100w/ miner_only_borrow_user_count * (block.number - startBlock) / 6 month;
   
- miner payback once: 
   60w/miner_payback_once_user_count * (block.number- startBlock) / 6 month;
   
- miner borrow and payback twice: 
   40w/miner_borrow_and_payback_ twice_user_count * (block.number- startBlock) / 6 month;
   
   
  ------------------------------------------------------------------------------------------------------ 
  If any user is missed, the administrator will transfer them offline using the batchTransfer contract
  
  - only manager can transfer token
  - only owner can add/delete manager
  - manager should approve the contract with enough token before transfer to users
  - manager and owner can't be the same address

### CODE REVIEW

**Review by zhiqiangxu — APPROVED** [review]
LGTM

File: contracts/Claim.sol [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    uint constant MINER_PAYBACK_SUPPLY = 60 * SUPPLY_UNIT;
+    uint constant MINER_PAYBACK_TWICE_SUPPLY = 40 * SUPPLY_UNIT;
+
+    // borrow airdrop will last 6 months
+    uint _startBlock = 0;
+    uint _releaseEndBlock = 0;
+    uint _airdropEndBlock = 0;
+    uint constant MONTH_IN_SECONDS = 30 * 86400;
+    uint constant BLOCK_SECONDS = 30;
+    uint constant RELEASE_LAST_TIME = 6 * MONTH_IN_SECONDS / BLOCK_SECONDS;                     // 6 months
```
[the remark this answers is not in the document]
Comment (dylenfu) [reply contracts/Claim.sol]: updated

File: contracts/Claim.sol [hunk 2]
```diff
@@ -208,30 +208,31 @@ contract Claim is Context {
     // 
     //-------------------------------------------------------------------------
     function calculateStake(address account, Request[] calldata requests) private view returns (CalculateResult memory r) {
-        bool[] memory actRecord = new bool[](uint(Action.ActionEnd));
+        bool[] memory actDumpRrcord = new bool[](uint(Action.ActionEnd));
```
Comment (zhiqiangxu) [inline contracts/Claim.sol:210]: `actDumpRrcord` should be `actDumpRecord`?

File: contracts/Claim.sol [hunk 3]
```diff
@@ -54,13 +54,24 @@ contract Claim is Context {
     uint private CLAIMABLE_ACTIONS = uint(Action.ActionEnd) - 1;
     uint _supplySum = FIG_BALANCE_SUPPLY + FIT_GOVERNANCE_SUPPLY + DISCORD_PHARSE2_SUPPLY + DISCORD_LEVEL5_SUPPLY + MINER_BORROW_SUPPLY + MINER_PAYBACK_SUPPLY + MINER_PAYBACK_TWICE_SUPPLY;
 
+    uint private calculateActionLength = 4;
```
Comment (zhiqiangxu) [inline contracts/Claim.sol:56]: where's this variable used?

Comment (dylenfu) [reply contracts/Claim.sol]: deleted

**Pull request merged**